python/ray/tests/test_object_store_metrics.py

```diff
@@ -0,0 +1,176 @@
+from collections import defaultdict
+import pytest
+from typing import Dict
+import numpy as np
+
+import ray
+from ray._private.test_utils import (
+    raw_metrics,
+    wait_for_condition,
+)
+from ray._private.worker import RayContext
+
+KiB = 1 << 10
+MiB = 1 << 20
+
+_SYSTEM_CONFIG = {
+    "automatic_object_spilling_enabled": True,
+    "max_io_workers": 100,
+    "min_spilling_size": 1,
+    "object_spilling_threshold": 0.99,  # to prevent premature spilling
+    "metrics_report_interval_ms": 100,
+}
+
+
+def objects_by_loc(info: RayContext) -> Dict:
+    res = raw_metrics(info)
+    objects_info = defaultdict(int)
+    if "ray_object_store_memory" in res:
+        for sample in res["ray_object_store_memory"]:
+            objects_info[sample.labels["Location"]] += sample.value
+
+    print(f"Objects by location: {objects_info}")
+    return objects_info
+
+
+def approx_eq_dict_in(actual: Dict, expected: Dict, e: int) -> bool:
+    """Check if two dict are approximately similar (with error allowed)"""
+    assert set(actual.keys()) == set(expected.keys()), "Unequal key sets."
+
+    for k, actual_v in actual.items():
+        expect_v = expected[k]
+        assert (
+            abs(expect_v - actual_v) <= e
+        ), f"expect={expect_v}, actual={actual_v}, diff allowed={e}"
+
+    return True
+
+
+def test_all_shared_memory(shutdown_only):
+    """Test objects allocated in shared memory"""
+    import numpy as np
+
+    info = ray.init(
+        object_store_memory=100 * MiB,
+        _system_config={
+            "metrics_report_interval_ms": 100,
+        },
+    )
+
+    # Allocate 80MiB data
+    objs_in_use = ray.get(
+        [ray.put(np.zeros(20 * MiB, dtype=np.uint8)) for _ in range(4)]
+    )
+
+    expected = {
+        "IN_MEMORY": 80 * MiB,
+        "SPILLED": 0,
+        "UNSEALED": 0,
+    }
+
+    wait_for_condition(
+        # 1KiB for metadata difference
+        lambda: approx_eq_dict_in(objects_by_loc(info), expected, 1 * KiB),
+        timeout=20,
+        retry_interval_ms=500,
+    )
+
+    # Free all of them
+    del objs_in_use
+
+    expected = {
+        "IN_MEMORY": 0,
+        "SPILLED": 0,
+        "UNSEALED": 0,
+    }
+
+    wait_for_condition(
+        # 1KiB for metadata difference
+        lambda: approx_eq_dict_in(objects_by_loc(info), expected, 1 * KiB),
+        timeout=20,
+        retry_interval_ms=500,
+    )
+
+
+def test_spilling(object_spilling_config, shutdown_only):
```

Comment:
Any idea on how to test the unsealed memory tracking specifically? 
It seems to me the create then seal path only happens when we restore spilled objects from storage directly. So one potential way I could test it is by mocking an external system with slow data copy. 

But I also feel that if the in memory tracking works properly, the unseal tracking should be the same since it's really similar code paths.

Reply:
it's probably tricky to do e2e. Maybe we can do at unit test level in cpp?

Reply:
So for cpp we do have unit test covering the unsealed stats correctness: https://github.com/ray-project/ray/blob/e9755d87a641b02daf82604ef23ce09647158cd4/src/ray/object_manager/plasma/test/stats_collector_test.cc#L75


Which is why I feel specifically testing the unsealed metrics is not super necessary: 
-  the tracking is tested correctly in c++
- the reporting is tested with the in memory memory testing e2e.